apps/api/src/withdrawal/withdrawal.service.ts

```diff
@@ -30,16 +49,70 @@ export class WithdrawalService {
     return result
   }
 
+  /**
+   * Updates a withdrawal, where status changes to completed/declined state will finilize the withdrawal and perform vault transaction.
+   */
   async update(id: string, updateWithdrawalDto: UpdateWithdrawalDto): Promise<Withdrawal | null> {
-    const result = await this.prisma.withdrawal.update({
+    const withdrawal = await this.prisma.withdrawal.findFirst({
+      where: { id: id },
+      rejectOnNotFound: true,
+    })
+
+    if ([WithdrawStatus.succeeded.valueOf(), WithdrawStatus.cancelled.valueOf(), WithdrawStatus.declined.valueOf()].includes(withdrawal.status.valueOf())) {
+      throw new BadRequestException("Withdrawal has already been finilized and cannot be updated.")
+    }
+    if (withdrawal.sourceVaultId !== updateWithdrawalDto.sourceVaultId || withdrawal.amount !== updateWithdrawalDto.amount) {
+      throw new BadRequestException("Vault or amount cannot be changed, please decline the withdrawal instead.")
+    }
+
+    const vault = await this.prisma.vault.findFirst({
+      where: {
+        id: withdrawal.sourceVaultId,
+      },
+      rejectOnNotFound: true,
+    })
+
+    // TODO: figure out how to initialize empty vault promise
+    let writeVault = this.prisma.vault.update({
+      where: { id: vault.id },
+      data: vault,
+    })
+    // in case of completion: complete transaction, unblock and debit the amount
+    if (updateWithdrawalDto.status === WithdrawStatus.succeeded) {
+      if (!updateWithdrawalDto.approvedById) {
+        throw new BadRequestException("Withdrawal needs to be approved by an authorized person.")
+      }
+      writeVault = this.prisma.vault.update({
+        where: { id: vault.id },
+        data: {
+          blockedAmount: {decrement: withdrawal.amount},
+          amount: {decrement: withdrawal.amount},
+        },
+      })
+    } else if (
+      updateWithdrawalDto.status === WithdrawStatus.declined ||
+      updateWithdrawalDto.status === WithdrawStatus.cancelled
+    ) {
+      // in case of rejection: unblock amount
+      writeVault = this.prisma.vault.update({
+        where: { id: vault.id },
+        data: { blockedAmount: {decrement: withdrawal.amount} },
+      })
+    }
```

Comment:
to avoid the if, can initialize above with the blockedAmount decrement

Reply:
same comment as before...